webrtc/dtls/dtls_connection.nim

```diff
@@ -0,0 +1,288 @@
+# Nim-WebRTC
+# Copyright (c) 2024 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+import chronos, chronicles
+import ../errors, ../stun/[stun_connection]
+import ./dtls_utils
+
+import mbedtls/ssl
+import mbedtls/ssl_cookie
+import mbedtls/ssl_cache
+import mbedtls/pk
+import mbedtls/md
+import mbedtls/ctr_drbg
+import mbedtls/rsa
+import mbedtls/x509
+import mbedtls/x509_crt
+import mbedtls/bignum
+import mbedtls/error
+import mbedtls/net_sockets
+import mbedtls/timing
+
+logScope:
+  topics = "webrtc dtls_conn"
+
+const DtlsConnTracker* = "webrtc.dtls.conn"
+
+type
+  MbedTLSCtx = object
+    ssl: mbedtls_ssl_context
+    config: mbedtls_ssl_config
+    cookie: mbedtls_ssl_cookie_ctx
+    cache: mbedtls_ssl_cache_context
+    timer: mbedtls_timing_delay_context
+    pkey: mbedtls_pk_context
+    srvcert: mbedtls_x509_crt
+    ctr_drbg: mbedtls_ctr_drbg_context
+
+  DtlsConn* = ref object
+    # DtlsConn is a Dtls connection receiving and sending data using
+    # the underlying Stun Connection
+    conn*: StunConn # The wrapper protocol Stun Connection
+    laddr: TransportAddress # Local address
+    raddr*: TransportAddress # Remote address
+    dataRecv: seq[byte] # data received which will be read by SCTP
+    sendFuture: Future[void].Raising([CancelledError, WebRtcError])
```

Comment:
Could you please elaborate more on why this is necessary?

Reply:
It's necessary because this Future is set during an Mbed-TLS callback.
Those callbacks are called by the `C` library thus cannot be asynchronous.
It basically goes this way https://github.com/vacp2p/nim-webrtc/pull/10/files#diff-33ff420ddc6f5b3b2a85d5b5b37dd7947486d2566811812f1183b721dc436852R231-R234